### PR TITLE
Add test that checks for tax expenditure differences

### PR DESCRIPTION
**Pull request by martinholmer**
In addition to the new test, also add a `tax-expenditures-report` target to Makefile.

Neither the new test nor the new Makefile target change any data-preparation code.

But this pull request also sets random number seeds used by the torch.Adam optimizer.
This was not done previously, so different random number streams were used in each Adam optimization, which uses **stochastic** gradient descent methods.  This meant the reweighting results were slightly different each time the reweighting was done.  This lack of reproducibility was the cause of different tax expenditure results on different computers described in [this comment](https://github.com/PSLmodels/tax-microdata-benchmarking/pull/155#issuecomment-2241256184).

As a results of adding the reweighting reproducibility, there are changes in the tax expenditure results.   The new results are as follows:
```
(taxcalc-dev) tax-microdata-benchmarking% cat tax_microdata_benchmarking/examination/tax_expenditures 
YR,KIND,EST= 2023 iitax 2212.0
YR,KIND,EST= 2023 ctc 129.2
YR,KIND,EST= 2023 eitc 75.8
YR,KIND,EST= 2023 social_security_partial_taxability 53.3
YR,KIND,EST= 2023 niit -57.0
YR,KIND,EST= 2023 cgqd_tax_preference 234.3
YR,KIND,EST= 2023 qbid 55.3
YR,KIND,EST= 2023 salt 20.8
YR,KIND,EST= 2026 iitax 2786.1
YR,KIND,EST= 2026 ctc 46.8
YR,KIND,EST= 2026 eitc 84.8
YR,KIND,EST= 2026 social_security_partial_taxability 72.2
YR,KIND,EST= 2026 niit -55.1
YR,KIND,EST= 2026 cgqd_tax_preference 232.5
YR,KIND,EST= 2026 qbid -0.0
YR,KIND,EST= 2026 salt 162.3
```
Which compare with the old PR #154 results:
```
< YR,KIND,EST= 2023 iitax 2212.7
< YR,KIND,EST= 2023 ctc 129.3
< YR,KIND,EST= 2023 eitc 75.8
< YR,KIND,EST= 2023 social_security_partial_taxability 53.3
< YR,KIND,EST= 2023 niit -57.0
< YR,KIND,EST= 2023 cgqd_tax_preference 234.3
< YR,KIND,EST= 2023 qbid 55.3
< YR,KIND,EST= 2023 salt 20.8
< YR,KIND,EST= 2026 iitax 2786.9
< YR,KIND,EST= 2026 ctc 46.8
< YR,KIND,EST= 2026 eitc 84.8
< YR,KIND,EST= 2026 social_security_partial_taxability 72.2
< YR,KIND,EST= 2026 niit -55.1
< YR,KIND,EST= 2026 cgqd_tax_preference 232.6
< YR,KIND,EST= 2026 qbid -0.0
< YR,KIND,EST= 2026 salt 162.4
```

### CODE REVIEW

**Comment by martinholmer** [comment]
Adding a test that checks if actual tax expenditures are exactly the same as expected tax expenditures will fail because the TMD files used to estimate the tax expenditures apparently differ on different computers.  For example, the tax expenditures that @donboyd5 generated on his computer differ slightly from the tax expenditures that @martinholmer generated on his computer.   (I have no idea why this is so, except for an unsubstantiated feeling that the rewieghting optimization may not be reaching a sensible convergence, and hence, leaving the "optimized" weights being slightly different on different computers.)

I'm changing the expected `examination/tax_expenditures` to be what they were on the GitHub run to see if a different GitHub computer generates the same results two times in a row.

**Comment by donboyd5** [comment]
Good news/bad news: the good news is that pr155 produces excellent results (as before) in terms of the many input variables I look at; the bad news is that on my machine it produces the same tax-expenditure results as before, which are slightly different from those produced by github. Here are the differences, slightly edited for readability. The most notable is iitax in 2023, where GH is about 0.03% higher than my results. The next two comments give (1) a follow-up question I put to ChatGPT and its response, and (2) my thoughts.

diff tax_microdata_benchmarking/storage/output/tax_expenditures \
             tax_microdata_benchmarking/examination/tax_expenditures

< YR,KIND,EST= 2023 iitax 2212.7
> YR,KIND,EST= 2023 iitax 2212.0

< YR,KIND,EST= 2023 ctc 129.3
> YR,KIND,EST= 2023 ctc 129.2

< YR,KIND,EST= 2026 iitax 2786.9
> YR,KIND,EST= 2026 iitax 2786.1

< YR,KIND,EST= 2026 cgqd_tax_preference 232.6
> YR,KIND,EST= 2026 cgqd_tax_preference 232.5

< YR,KIND,EST= 2026 salt 162.4
> YR,KIND,EST= 2026 salt 162.3
make: *** [Makefile:26: tax-expenditures-report] Error 1

**Comment by donboyd5** [comment]
Here's a question I put to ChatGPT and its response:


### Question to ChatGPT:
I am running an optimizer in python using pytorch and Adam, which I believe uses stochastic gradient descent (is this correct?) I am getting very slightly different results on different computers, although repeated results on either computer individually appear to be identical. I would like to have 100% reproducible results that are the same on both machines.

I've copied an excerpt of the code (slightly simplified) below. Can you tell me why the results might not be identical across machines and what I can do to make them so?

  ```
rng_seed = 65748392
  torch.manual_seed(rng_seed)  # set the random number seed for CPU
  torch.cuda.manual_seed_all(rng_seed)  # set the seed for all GPUs
  optimizer = torch.optim.Adam([weight_multiplier], lr=1e-1)

 for i in tqdm(range(2_000), desc="Optimising weights"):
        optimizer.zero_grad()
        new_weights = weights * (
            torch.clamp(
                weight_multiplier,
                min=weight_multiplier_min,
                max=weight_multiplier_max,
            )
        )
        outputs = (new_weights * output_matrix_tensor.T).sum(axis=1)
        loss_value = (
            ((outputs + 1) / (target_array + 1) - 1) ** 2
        ).sum() + weight_deviation
        loss_value.backward()
        optimizer.step()
```


### ChatGPT response:
The slight differences in results you are observing on different machines when running an optimizer in PyTorch, such as Adam (which is indeed a type of stochastic gradient descent), are a common issue in deep learning environments due to several factors:

**Factors Contributing to Non-Reproducibility**
**_Different Hardware_**: Slight differences in computations can occur due to variations in hardware, especially between CPUs and GPUs, or different models/brands of GPUs. These variations are often due to how floating-point calculations are handled by different processors.

**_Parallelism_**: Non-deterministic behavior can arise from parallel computing elements in GPUs and even CPUs. This is because reductions and other parallel operations do not guarantee the order of operations, which can lead to tiny variations in results.

**_Library Implementations_**: Variations in the backend libraries across platforms (e.g., different versions of CUDA or cuDNN) can lead to slight differences in outcomes.

**Steps to Ensure Reproducibility**
To maximize reproducibility across different machines, consider implementing the following in your PyTorch code:

**_Set All Random Seeds_**: You've already set the seeds for CPU and all GPUs, which is great. Make sure that all parts of your code that rely on random number generation are controlled.

**_Control Sources of Non-Determinism_**:

If you are using GPUs, ensure that you set PyTorch to work in a deterministic manner as much as possible. You can do this by:
python

```
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

This can potentially reduce performance due to the disabling of certain optimizations, but it increases reproducibility.
Use the Same Hardware and Software Environments:

Where possible, use the same type of hardware and software versions across different machines. This includes the same GPU models, the same versions of PyTorch, CUDA, and cuDNN.

**Control Floating-Point Precision**:

Consider the precision of floating-point calculations. Using double precision (float64) instead of the typical single precision (float32) can reduce discrepancies but at the cost of increased computation and memory usage.
You can set tensors to use double precision by default with:

`torch.set_default_dtype(torch.float64)`

**Review the Optimizer and Mathematical Operations:**

Check that the optimizer and any custom operations are implemented in a way that doesn't introduce hardware-dependent variations. For example, certain operations might behave slightly differently on different GPUs.

**_Version Control for Libraries_**:

Ensure that all external libraries and dependencies are the same version across all machines. This includes the versions of PyTorch, Python, and any related libraries.

**Final Considerations**

Achieving 100% reproducibility in deep learning across different hardware and software environments can be challenging and sometimes not feasible due to inherent hardware differences and external dependencies. However, by controlling as many of the above factors as possible, you can minimize these differences significantly. If exact reproducibility is critical, consider running your experiments on the same machine or set of machines with controlled environments.

**Comment by donboyd5** [comment]
Realistically, reading the ChatGPT response in combination with my prior investigation into parallelism questions, I think we should accept that high-end optimizers may produce very slightly different results on different equipment and perhaps even on the same equipment. It seems like people doing this work generally accept it. The issues for us should be (1) making sure that variations are acceptably small, and (2) making sure our tests don't fail solely because we have differences that are acceptably small.

**Making the differences acceptably small**
Arguably, the $0.8 billion 0.03% difference in iitax in 2026 is acceptably small (as is the $0.1 billion 0.08% difference in CTC in 2023), although I wouldn't mind trying to get to the point where numbers are identical to 5 significant digits, which they are not (and sometimes not to 4 significant digits). However, on my machine (and perhaps on github) all of the 500+ targeted values that I have looked at are certainly close enough to targets to be excellent for policy analysis. We are so close to targets, and so much closer to targets than taxdata is, that I think this worry is a 3rd or 4th order issue.

I am not familiar with pytorch and Adam except for a very little bit of reading on it, and I don't think we should embark on a significant project to make the differences smaller than they are now, particularly at this stage. I have always thought that our approach to optimization here could be sped up and streamlined. In my past work, I have used other optimizers that are deterministic and have solved problems of this structure and size in a fraction of the time that Adam does. So if we were to put a lot of effort into improving optimization, I would want to explore that. But Adam has worked well (and continues to work well) and delving into this was an unaffordable luxury that we couldn't work on; maybe for a future phase if there is one.

So if we want to explore this, I think we should only do a little bit. According to a follow-up conversation I had with ChatGPT, decreasing the learning rate or increasing the number of iterations both could increase the precision of results (and there are other ways, too). I am currently running pr155 with the learning rate reduced from 1e-1 to 1e-3. I'll post the tax-expenditure results shortly, along with my subjective assessment of target-hitting accuracy. Perhaps you could then see if, making the same change, you get tax-expenditure results that are closer to my lr=1e-3 than when we both used lr=1e-1?

**Not failing tests solely because of acceptably small differences**

Unless lr=1e-3 makes differences go away at 5 significant digits, I suggest we simply increase the tolerances around our tax-expenditure tests. For example, perhaps we should consider them passing if we are within 0.5% of GH results.

**Comment by donboyd5** [comment]
Setting lr=1e-3 is dramatically worse.

Here are the tax expenditure results:

diff tax_microdata_benchmarking/storage/output/tax_expenditures \
             tax_microdata_benchmarking/examination/tax_expenditures
1,14c1,14
< YR,KIND,EST= 2023 iitax 2200.3
< YR,KIND,EST= 2023 ctc 128.8
< YR,KIND,EST= 2023 eitc 76.0
< YR,KIND,EST= 2023 social_security_partial_taxability 53.4
< YR,KIND,EST= 2023 niit -51.6
< YR,KIND,EST= 2023 cgqd_tax_preference 200.3
< YR,KIND,EST= 2023 qbid 56.9
< YR,KIND,EST= 2023 salt 21.1
< YR,KIND,EST= 2026 iitax 2771.2
< YR,KIND,EST= 2026 ctc 46.6
< YR,KIND,EST= 2026 eitc 85.0
< YR,KIND,EST= 2026 social_security_partial_taxability 72.3
< YR,KIND,EST= 2026 niit -49.3
< YR,KIND,EST= 2026 cgqd_tax_preference 194.4
---
> YR,KIND,EST= 2023 iitax 2212.0
> YR,KIND,EST= 2023 ctc 129.2
> YR,KIND,EST= 2023 eitc 75.8
> YR,KIND,EST= 2023 social_security_partial_taxability 53.3
> YR,KIND,EST= 2023 niit -57.0
> YR,KIND,EST= 2023 cgqd_tax_preference 234.3
> YR,KIND,EST= 2023 qbid 55.3
> YR,KIND,EST= 2023 salt 20.8
> YR,KIND,EST= 2026 iitax 2786.1
> YR,KIND,EST= 2026 ctc 46.8
> YR,KIND,EST= 2026 eitc 84.8
> YR,KIND,EST= 2026 social_security_partial_taxability 72.2
> YR,KIND,EST= 2026 niit -55.1
> YR,KIND,EST= 2026 cgqd_tax_preference 232.5
16c16
< YR,KIND,EST= 2026 salt 160.8
---
> YR,KIND,EST= 2026 salt 162.3
make: *** [Makefile:26: tax-expenditures-report] Error 1

**Comment by martinholmer** [comment]
@donboyd5 said in PR #155:
> Arguably, the $0.8 billion 0.03% difference in iitax in 2026 is acceptably small (as is the $0.1 billion 0.08% difference in CTC in 2023), although I wouldn't mind trying to get to the point where numbers are identical to 5 significant digits, which they are not (and sometimes not to 4 significant digits). However, on my machine (and perhaps on github) all of the 500+ targeted values that I have looked at are certainly close enough to targets to be excellent for policy analysis. We are so close to targets, and so much closer to targets than taxdata is, that I think this worry is a 3rd or 4th order issue.
> ...
> So if we want to explore this [making reweighting more accurate], I think we should only do a little bit.

I'm in agreement with the "little bit" part, where "little bit" includes nothing.
